src/pybind/mgr/dashboard/frontend/src/app/shared/services/notification.service.ts

```diff
@@ -87,41 +88,62 @@ export class NotificationService {
    * @param {string} [message] The message to be displayed. Note, use this field
    *   for error notifications only.
    * @param {*} [options] toastr compatible options, used when creating a toastr
+   * @param {string} [application] Only needed if notification comes from an external application
    * @returns The timeout ID that is set to be able to cancel the notification.
    */
-  show(type: NotificationType, title: string, message?: string, options?: any): number;
-  show(config: CdNotificationConfig): number;
   show(
-    arg: NotificationType | CdNotificationConfig,
+    type: NotificationType,
+    title: string,
+    message?: string,
+    options?: any | ToastOptions,
+    application?: string
+  ): number;
+  show(config: CdNotificationConfig | (() => CdNotificationConfig)): number;
+  show(
+    arg: NotificationType | CdNotificationConfig | (() => CdNotificationConfig),
     title?: string,
     message?: string,
-    options?: any
+    options?: any | ToastOptions,
+    application?: string
   ): number {
-    let type;
-    if (_.isObject(arg)) {
-      ({ message, type, title, options } = <CdNotificationConfig>arg);
-    } else {
-      type = arg;
-    }
     return window.setTimeout(() => {
-      this.save(type, title, message);
-      if (!message) {
-        message = '';
-      }
-      switch (type) {
-        case NotificationType.error:
-          this.toastr.error(message, title, options);
-          break;
-        case NotificationType.info:
-          this.toastr.info(message, title, options);
-          break;
-        case NotificationType.success:
-          this.toastr.success(message, title, options);
-          break;
+      let config: CdNotificationConfig;
+      if (_.isFunction(arg)) {
+        config = <CdNotificationConfig>arg();
+      } else if (_.isObject(arg)) {
+        config = <CdNotificationConfig>arg;
+      } else {
+        config = new CdNotificationConfig(
+          <NotificationType>arg,
+          title,
+          message,
+          options,
+          application
+        );
       }
+      const notification = new CdNotification(config);
+      this.save(notification);
+      this.showToasty(notification);
     }, 10);
   }
 
+  private showToasty(notification: CdNotification) {
+    this.toastr[['error', 'info', 'success'][notification.type]](
+      (notification.message ? notification.message + '<br>' : '') +
+        this.getTimeAndApplicationHtml(notification),
+      notification.title,
+      notification.options
+    );
+  }
+
+  getTimeAndApplicationHtml(notification: CdNotification): string {
```

Comment:
```suggestion
  renderTimeAndApplication(notification: CdNotification): string {
```
or
```suggestion
  renderTimeAndApplicationHtml(notification: CdNotification): string {
```
To make it more clear what this method is doing.